### PR TITLE
Add nccl package and Bump vLLM to 0.4.3 for huggingface runtime

### LINKED ISSUE #3722: `NCCL` and `flash_atten`  packages are missing in huggingface runtime
/kind bug

**What steps did you take and what happened:**
```
INFO 06-03 18:32:32 pynccl_utils.py:17] Failed to import NCCL library: Cannot find libnccl.so.2 in the system.
INFO 06-03 18:32:32 pynccl_utils.py:18] It is expected if you are not running on NVIDIA GPUs.
INFO 06-03 18:32:32 selector.py:81] Cannot use FlashAttention-2 backend because the flash_attn package is not found. Please install it for better performance. 
```
it seems NCCL library and flash_attn package are missing. kserve-container seems to be running this image id [docker.io/kserve/huggingfaceserver@sha256:a3375bcece636afae2ec11928c4fc295d712e00e54e0e69908f978ee543ba6f3](http://docker.io/kserve/huggingfaceserver@sha256:a3375bcece636afae2ec11928c4fc295d712e00e54e0e69908f978ee543ba6f3)
Here is the full log
```
2024-06-03 18:32:31.292 1 kserve INFO [storage.py:download():66] Copying contents of /mnt/models/ to local
INFO 06-03 18:32:31 llm_engine.py:100] Initializing an LLM engine (v0.4.2) with config: model='/mnt/models/', speculative_config=None, tokenizer='/mnt/models/', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), seed=0, served_model_name=/mnt/models/)
INFO 06-03 18:32:32 pynccl_utils.py:17] Failed to import NCCL library: Cannot find libnccl.so.2 in the system.
INFO 06-03 18:32:32 pynccl_utils.py:18] It is expected if you are not running on NVIDIA GPUs.
INFO 06-03 18:32:32 selector.py:81] Cannot use FlashAttention-2 backend because the flash_attn package is not found. Please install it for better performance.
INFO 06-03 18:32:32 selector.py:32] Using XFormers backend.
INFO 06-03 18:33:03 model_runner.py:175] Loading model weights took 12.5523 GB
INFO 06-03 18:33:04 gpu_executor.py:114] # GPU blocks: 826, # CPU blocks: 512
INFO 06-03 18:33:06 model_runner.py:937] Capturing the model for CUDA graphs. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 06-03 18:33:06 model_runner.py:941] CUDA graphs can take additional 1~3 GiB memory per GPU. If you are running out of memory, consider decreasing `gpu_memory_utilization` or enforcing eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 06-03 18:33:15 model_runner.py:1017] Graph capturing finished in 9 secs.
2024-06-03 18:33:15.381 1 kserve INFO [model_server.py:register_model():382] Registering model: kserve-native-vllm-llama-2-7b
2024-06-03 18:33:15.382 1 kserve INFO [model_server.py:start():252] Setting max asyncio worker threads as 5
2024-06-03 18:33:15.382 1 kserve INFO [model_server.py:serve():258] Starting uvicorn with 1 workers
2024-06-03 18:33:15.510 uvicorn.error INFO:     Started server process [1]
2024-06-03 18:33:15.510 uvicorn.error INFO:     Waiting for application startup.
2024-06-03 18:33:15.537 1 kserve INFO [server.py:start():63] Starting gRPC server on [::]:8081
2024-06-03 18:33:15.537 uvicorn.error INFO:     Application startup complete.
2024-06-03 18:33:15.537 uvicorn.error INFO:     Uvicorn running on http://0.0.0.0:8080/ (Press CTRL+C to quit)
```

**What did you expect to happen:**

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: 0.13.0rc1
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
- From 0.4.3 vLLM uses [`vllm-flash-attn`](https://github.com/vllm-project/vllm/pull/4686) instead of `flash-attn` package which is specified as a required dependency.
- Updates the huggingface runtime base image to `nvidia/cuda:12.1.0-cudnn8-runtime-ubuntu22.04` which has the `nccl` package pre-installed. 
- Updates CUDA version to 12.4

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
- Fixes #3722

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Test A
```
$ (.venv) root@768af757c26b:/workspace/kserve/python# python3.10 -m huggingfaceserver --model_id=facebook/opt-125m --model_name=opt-125m
/workspace/kserve/.venv/lib/python3.10/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
INFO 06-04 10:46:41 llm_engine.py:161] Initializing an LLM engine (v0.4.3) with config: model='facebook/opt-125m', speculative_config=None, tokenizer='facebook/opt-125m', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, rope_scaling=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=2048, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), seed=0, served_model_name=facebook/opt-125m)
INFO 06-04 10:46:42 weight_utils.py:207] Using model weights format ['*.bin']
INFO 06-04 10:46:43 model_runner.py:146] Loading model weights took 0.2389 GB
INFO 06-04 10:46:43 gpu_executor.py:83] # GPU blocks: 37210, # CPU blocks: 7281
INFO 06-04 10:46:45 model_runner.py:854] Capturing the model for CUDA graphs. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 06-04 10:46:45 model_runner.py:858] CUDA graphs can take additional 1~3 GiB memory per GPU. If you are running out of memory, consider decreasing `gpu_memory_utilization` or enforcing eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 06-04 10:46:51 model_runner.py:924] Graph capturing finished in 5 secs.
2024-06-04 10:46:51.555 6696 kserve INFO [model_server.py:register_model():384] Registering model: opt-125m
2024-06-04 10:46:51.556 6696 kserve INFO [model_server.py:start():254] Setting max asyncio worker threads as 11
2024-06-04 10:46:51.556 6696 kserve INFO [model_server.py:serve():260] Starting uvicorn with 1 workers
2024-06-04 10:46:51.663 uvicorn.error INFO:     Started server process [6696]
2024-06-04 10:46:51.663 uvicorn.error INFO:     Waiting for application startup.
2024-06-04 10:46:51.667 6696 kserve INFO [server.py:start():63] Starting gRPC server on [::]:8081
2024-06-04 10:46:51.667 uvicorn.error INFO:     Application startup complete.
2024-06-04 10:46:51.668 uvicorn.error INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
```

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add nccl package and Bump vLLM to 0.4.3 for huggingface runtime to fix flash_attention package not available error.
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/huggingface_server.Dockerfile [hunk 1]
```diff
@@ -34,7 +34,7 @@ RUN cd huggingfaceserver && poetry install --no-interaction --no-cache
 
 RUN pip3 install vllm==${VLLM_VERSION}
 
-FROM nvidia/cuda:12.1.0-base-ubuntu22.04 as prod
+FROM nvidia/cuda:12.1.0-cudnn8-runtime-ubuntu22.04 as prod
```
Comment (sivanantha321) [inline python/huggingface_server.Dockerfile]: @yuzisun Do we need cuDNN ?

**Comment by Datta0** [comment]
Also with this change to [vLLM](https://github.com/vllm-project/vllm/commit/5ae5ed1e6047d4095149e26526a618be0529a118#diff-4bd825485b5162cf8021da41a2ebd3d4026929e617d1137f69bbbe4bda0ee643L249-R251) there's no more prompt_token_ids
So [this](https://github.com/kserve/kserve/blob/ff744c60f3189224e5724b7b2fdd6ebf567db327/python/huggingfaceserver/huggingfaceserver/vllm/vllm_completions.py#L150) would break in kserve

We should fix this before updating vLLM to 0.4.3?

**Comment by sivanantha321** [comment]
> Also with this change to [vLLM](https://github.com/vllm-project/vllm/commit/5ae5ed1e6047d4095149e26526a618be0529a118#diff-4bd825485b5162cf8021da41a2ebd3d4026929e617d1137f69bbbe4bda0ee643L249-R251) there's no more prompt_token_ids So [this](https://github.com/kserve/kserve/blob/ff744c60f3189224e5724b7b2fdd6ebf567db327/python/huggingfaceserver/huggingfaceserver/vllm/vllm_completions.py#L150) would break in kserve
> 
> We should fix this before updating vLLM to 0.4.3?

Good Catch! Fixed

File: python/huggingface_server.Dockerfile [hunk 2]
```diff
@@ -9,7 +9,7 @@ ARG POETRY_HOME=/opt/poetry
 ARG POETRY_VERSION=1.7.1
 
 # Install vllm
-ARG VLLM_VERSION=0.4.2
+ARG VLLM_VERSION=0.4.3
```
Comment (yuzisun) [inline python/huggingface_server.Dockerfile:12]: Any reason to upgrade to 0.4.3 ? seems like there are some breaking changes which affects us.

File: python/huggingface_server.Dockerfile (hunk 1 above)
Comment (yuzisun) [reply python/huggingface_server.Dockerfile]: is this needed for nccl installation ?

**Comment by datta-nimmaturi** [comment]
From my brief testing/benchmarking, I don't see any significant improvement either by `flash-attn` or by `vllm-flash-attn` over the default. They're within the 1% range of each other. 
And both with or without flash-attn, vLLM was able to create same number of blocks for paged attention aka same memory usage
[This](https://github.com/Dao-AILab/flash-attention/issues/427#issuecomment-1668257877) is what Tri Dao said about the same (old though)
[flash-decoding](https://pytorch.org/blog/flash-decoding/) which was added in FlashAttention 2.2 might help but I don't know where the said performance improvements are vanishing

FlashAttention as of now is a good to have but doesn't give us enough to necessitate it ig?

**Comment by Datta0** [comment]
From my brief testing/benchmarking, I don't see any significant improvement either by `flash-attn` or by `vllm-flash-attn` over the default. They're within the 1% range of each other. 
And both with or without flash-attn, vLLM was able to create same number of blocks for paged attention aka same memory usage
[This](https://github.com/Dao-AILab/flash-attention/issues/427#issuecomment-1668257877) is what Tri Dao said about the same (old though)
[flash-decoding](https://pytorch.org/blog/flash-decoding/) which was added in FlashAttention 2.2 might help but I don't know where the said performance improvements are vanishing

**Comment by sivanantha321** [comment]
> From my brief testing/benchmarking, I don't see any significant improvement either by flash-attn or by vllm-flash-attn over the default. They're within the 1% range of each other.
And both with or without flash-attn, vLLM was able to create same number of blocks for paged attention aka same memory usage
https://github.com/Dao-AILab/flash-attention/issues/427#issuecomment-1668257877 is what Tri Dao said about the same (old though)
[flash-decoding](https://pytorch.org/blog/flash-decoding/) which was added in FlashAttention 2.2 might help but I don't know where the said performance improvements are vanishing

vLLM does adds [support for decoding using `flash-attn`](https://github.com/vllm-project/vllm/pull/3648) in 0.4.3 and according to [this comment](https://github.com/vllm-project/vllm/pull/3648#issuecomment-2025545602) it gives roughly 10% increase in throughput. Btw, `vllm-flash-attn` and `flash-atten` are essentially the same. It is just a [work around for wheel size limitation.](https://github.com/vllm-project/vllm/pull/3648#issuecomment-2097150641). `Flash-atten` has some constraints on [the head size as explained here.](https://github.com/vllm-project/vllm/pull/3648#issuecomment-2120182846). So if the model does not falls under this head size, then it won't enjoy any performance improvements from `flash-atten`.

**Comment by sivanantha321** [comment]
I am just adding [a Note from vllm docs](https://github.com/vllm-project/vllm/blob/main/docs/source/serving/deploying_with_docker.rst) here:
For `v0.4.1` and `v0.4.2` only - the vLLM docker images under these versions are supposed to be run under the root user since a library under the root user's home directory, i.e. /root/.config/vllm/nccl/cu12/libnccl.so.2.18.1 is required to be loaded during runtime. If you are running the container under a different user, you may need to first change the permissions of the library (and all the parent directories) to allow the user to access it, then run vLLM with environment variable VLLM_NCCL_SO_PATH=/root/.config/vllm/nccl/cu12/libnccl.so.2.18.1 .

**Comment by sivanantha321** [comment]
> I am just adding [a Note from vllm docs](https://github.com/vllm-project/vllm/blob/main/docs/source/serving/deploying_with_docker.rst) here: For `v0.4.1` and `v0.4.2` only - the vLLM docker images under these versions are supposed to be run under the root user since a library under the root user's home directory, i.e. /root/.config/vllm/nccl/cu12/libnccl.so.2.18.1 is required to be loaded during runtime. If you are running the container under a different user, you may need to first change the permissions of the library (and all the parent directories) to allow the user to access it, then run vLLM with environment variable VLLM_NCCL_SO_PATH=/root/.config/vllm/nccl/cu12/libnccl.so.2.18.1 .

So to use NCCL we have to run vLLM as root user for 0.4.1 and 0.4.2. But that's not the case for 0.4.3

File: python/huggingface_server.Dockerfile (hunk 1 above)
Comment (sivanantha321) [reply python/huggingface_server.Dockerfile]: `nvidia/cuda:12.1.0-runtime-ubuntu22.04` image comes with NCCL pre-installed. If we don't need cuDNN, We can use this image.

**Comment by Datta0** [comment]
@sivanantha321 
I tried mistral as said in the [comment](https://github.com/vllm-project/vllm/pull/3648#issuecomment-2025545602) with and without flash attn
These are the results
mistral-7b-v0.1

withouth flash-attn
```
Avg latency: 1.8139188961683734 seconds
10% percentile latency: 1.8072806416093954 seconds
25% percentile latency: 1.8097923692512268 seconds
50% percentile latency: 1.813208505504008 seconds
75% percentile latency: 1.8156179094985418 seconds
90% percentile latency: 1.8222627914961778 seconds
```
```
Throughput: 9.91 requests/s, 4680.56 tokens/s
```

with flash-attn
```
Avg latency: 1.8172954017999776 seconds
10% percentile latency: 1.813928227701399 seconds
25% percentile latency: 1.81456515800528 seconds
50% percentile latency: 1.8163161535048857 seconds
75% percentile latency: 1.8180346899971482 seconds
90% percentile latency: 1.8244814682984725 seconds
```
```
Throughput: 10.36 requests/s, 4892.64 tokens/s
```

4% delta in throughput and <1% delta i latencies if my math serves me right

Maybe its more pronounced for larger models idk...

**Comment by johnugeorge** [comment]
Also, if it has to be run as a root user,  we have to make it optional as it might not be possible in all envs. @sivanantha321

**Comment by sivanantha321** [comment]
> Also, if it has to be run as a root user, we have to make it optional as it might not be possible in all envs. @sivanantha321

If we are upgrading to 0.4.3, this is not required. Should we stick to 0.4.2 for this release ?

**Comment by robertgshaw2-redhat** [comment]
Hey just curious - is there any reason why you could not start from the vLLM Docker image?

We are going to have some new features coming out which are going to require CUDA 12.4 (which is now what we use in the official vLLM docker image `FROM nvidia/cuda:12.4.1-base-ubuntu22.04 AS vllm-base`). (in particular, we have some cutlass kernels for fp8 on Ada lovelace, which need CUDA 12.4)

It might be easier to stay in sync with the vLLM project this way

**Comment by sivanantha321** [comment]
> Hey just curious - is there any reason why you could not start from the vLLM Docker image?
> 
> We are going to have some new features coming out which are going to require CUDA 12.4 (which is now what we use in the official vLLM docker image `FROM nvidia/cuda:12.4.1-base-ubuntu22.04 AS vllm-base`). (in particular, we have some cutlass kernels for fp8 on Ada lovelace, which need CUDA 12.4)
> 
> It might be easier to stay in sync with the vLLM project this way

This runtime uses `vLLm` as one of the backends. In future, We plan on adding some more backends as well. So, Using `cuda` image as base image gives more advantage than relying on vLLM docker image.

**Pull request merged**